Fault Tree Analysis & Constraint Diagrams

prmoko opened the pull request:
Created a +SafetyPkg for a Fault Tree Analysis and a +ConstraintDiagramPkg for generating constraint diagrams for aircraft certified under FAA Part 25 requirements. The Fault Tree Analysis code is compatible with GNU Octave.

prmoko committed:
Created FaultTree

Developed code to read a matrix, determine which nodes are sources/sinks, and print whether the node is a source, sink, or intermediate node.

prmoko committed:
Tree Plotted

Added the capability to traverse a propulsion architecture and create a tree indicating the possible failures that could occur. It is not yet a fault tree because AND/OR gates are missing.

prmoko committed:
Logic Gates in FaultTree

Added code to insert logic gates in the fault tree. Right now, any fault that is connected to one other component is assigned an 'OR' gate and any fault that is connected to multiple components is assigned an 'AND' gate.

prmoko committed:
Created CreateFaultTree

Wrote this code to create a fault tree. It is similar to the other code that was written previously, but the AND/OR gates are selected more formally. Also, including an example fault tree and the breadth-/depth-first searches used when traversing a graph.

prmoko committed:
Updated FaultTree

Updated code to check which logic gates are used in the tree such that the connections between failures and gates are performed dynamically. Previously, this was done without considering that some logic gates may go unused (for primary events).

prmoko committed:
Updated ExampleFT

Changed the component failure names to be more descriptive for plotting on a fault tree.

prmoko committed:
Updated FaultTree

Resolved issues with stashed and current versions.

prmoko committed:
Updated FaultTree

Corrected the node-gate connections that were previously erroneous. Instead of considering the number of edges in the graph and the order that the nodes are traversed, just allocate enough memory for all of the nodes (which include failures and logic gates).

prmoko committed:
Updated CreateFaultTree

Included downstream and internal failures, as well as assigning which nodes receive which type of logic gate.

prmoko committed:
Renamed FaultTree

Renamed 'FaultTree' to 'DrawFaultTree', which more accurately describes what it is doing.

prmoko committed:
Updated CreateFaultTree

Modified the function to call the new function that draws the fault tree.

prmoko committed:
Updated DrawFaultTree

Modified the doucmentation to reflect the correct inputs.

prmoko committed:
Updated CreateFaultTree

Removed extra lines of code that are no longer used.

prmoko committed:
Updated CreateFaultTree

If the downstream failure relies on a single component, just use that component instead of creating an additional OR gate to connect the downstream failure to the upstream component.

prmoko committed:
Updated CreateFaultTree

Simplified the logic for identifying internal failures.

prmoko committed:
Example Fault Tree Cases

Renamed the first fault tree example. Also, added a second fault tree example to test how shared primary events are handled in the fault tree analysis.

prmoko committed:
Updated CreateFaultTree

Began accounting for a single repeated primary event in the fault tree. This involves re-arranging the fault tree leaves. Multiple repeated primary events cannot be accounted for yet.

prmoko committed:
Added Fault Tree Examples

Created more fault tree examples to stress-test the code and make it more robust.

prmoko committed:
Updated CreateFaultTree

Began accounting for multiple 'layers' of redundant components, but they still need the appropriate logic gates to be added.

prmoko committed:
Updated CreateFaultTree

Added ability to account for multiple repeated primary events at one level of the fault tree. Additional debugging is needed to account for multiple repeated primary events at multiple levels of the fault tree.

prmoko committed:
Created ExampleFT06

Created a new fault tree test case that involves a portion of the SUSAN propulsion architecture.

prmoko committed:
Updated CreateFaultTree

Fixed code to account for multiple repeated primary events in a fault tree.

prmoko committed:
Added ExampleFT07

Added a fault tree example with redundancies that cannot be re-arranged in the fault tree.

prmoko committed:
Created AssessFaultTree

Initial approach to assessing a fault tree with Boolean logic rather than re-drawing the fault tree.

prmoko committed:
Updated AssessFaultTree

Improved the function such that it creates a string of failures based on the system architecture provided. Then, Matlab's symbolics are used to simplify the Boolean expression before substituting in the failure probabilities and computing the system's probability of failure.

prmoko committed:
Added ARP4761 Test Case

Added a simple test case from ARP4761A to check the new fault tree analysis.

prmoko committed:
Updated AssessFaultTree

After simplifying, expand one last time to more easily extract all unique failure modes.

prmoko committed:
Created FindMinimumCutSets

Began writing a function that finds the minimum cut sets for a given system architecture. Right now, the function only returns the cut sets and does not reduce it to the minimum cut set. Also, the probability that the system fails is not yet computed.

prmoko committed:
Updated FindMinimumCutSets

Added Boolean algebra simplification for repeated failures in a single failure mode. This represents the Idempotent Law of Boolean Algebra: X * X = X.

prmoko committed:
Updated FindMinimumCutSets

Added Boolean logic to eliminate common events between different failure modes. Represents X + X * Y = X in Boolean algebra.

prmoko committed:
Updated FindMinimumCutSets

Added ability to compute the system-level failure rate. The code was tested and validated against the ExampleFT**.mat files (failure rates were computed by hand).

prmoko committed:
Updated FindMinimumCutSets

Vectorized the Boolean algebra simplification of X * X = X.

prmoko committed:
Updated FindMinimumCutSets

Accounted for the fact that a component may not have an internal failure. In this case, the 'Failures' matrix will need an extra column of empty strings.

prmoko committed:
Updated FindMinimumCutSets

Took faster approach to enumerating combinations.

prmoko committed:
Small ways to speedup the code a little bit.

prmoko committed:
Updated FindMinimumCutSets

Added a second output to return all the failure modes for user-analysis and post-processing.

prmoko committed:
Updated FindMinimumCutSets

After enumerating all failures, reduce using the Imdepotent Law and Law of Absorbption at lower levels before returning to the higher level system.

prmoko committed:
File Operations

Deleted an old FTA analysis and renamed the existing one.

prmoko committed:
Created FailureModel

Created a function to use the CDF of an exponential distribution for computing the probability that a component fails.

prmoko committed:
Helper Functions for Reliability

Created a component database to keep track of failure rates per hour and a function to plot the probability that a component fails as a function of exposure time.

prmoko committed:
Updated FaultTreeAnalysis

Instead of finding the connections in the recursive loop, do it outside the loop ahead of time.

prmoko committed:
Updated FaultTreeAnalysis

Parallelized parts of the algorithm.

prmoko committed:
Updated FaultTreeAnalysis

Parallelized reductions.

prmoko committed:
Updated FaultTreeAnalysis

Attempted to parallelize.

prmoko committed:
Updated FaultTreeAnalysis

Attempting to reduce run time by simplifying after consecutive enumerations rather than waiting until the end.

prmoko committed:
Updated FaultTreeAnalysis

De-conflicted the two versions.

prmoko committed:
Updated FaultTreeAnalysis

Sped up Idempotent Law simplification by evaluating only the new columns that are added to the enumeration.

prmoko committed:
Updated FaultTreeAnalysis

Modified how singular failure modes are enumerated in the event that the bottom-level events are not to be considered in the FTA.

prmoko committed:
Updated Create/DrawFaultTree

Modified the functions such that multiple AND/OR gates can be connected consecutively and don't require extraneous intermediate events.

prmoko committed:
Updated FaultTreeAnalysis

Instead of checking for 1s only, check each column to determine how many inputs are needed to trigger the gate.

prmoko committed:
New FTA Test Cases

Created new FTA test cases where the matrix is no longer 1s and 0s, but has whole numbers in each entry containing a connection. In each column, replace all 1s with the number of components that must fail to trigger a gate.

prmoko committed:
Updated FaultTreeAnalysis

Added a k-out-of-n gate to the analysis, which goes along with the new matrices that define the number of inputs to be triggered for the gate to be flagged.

prmoko committed:
Updated FaultTreeAnalysis

Updated AND gate to only enumerate if there are two sets of failures. If one set of failures is provided, the gate exits without returning an enumerated set.

prmoko committed:
Updated FaultTreeAnalysis

Updated AND gate to only enumerate if all inputs are present.

prmoko committed:
Added ConstraintDiagramPkg

Transferred old constraint diagram code from another repository to this one.

prmoko committed:
Updated ConstraintDiagram

Modified the code to include constraints for sizing jet aircraft. Removed old/excess code and made modifications to how the plot is presented.

prmoko committed:
Constraints for Jet Aircraft

Created individual function files for each sizing constraint provided in Prof. Martins' Aircraft Design Metabook. Each 14 ECFR 25 regulation can be identified by its function name. Other functions relate to the service ceiling, cruise altitude, and takeoff/landing field lengths.

prmoko committed:
Created TestConstraintDiagram

Test the ConstraintDiagram function by modeling a constraint diagram for an Airbus A320.

prmoko committed:
Updated Jet25_111

Modified the parasite drag coefficient to exclude landing gear deployment (already retracted, according to 14 ECFR 121b.

prmoko committed:
Updated TestConstraintDiagram

Removed commented code that was unnecessary and made other small edits.

prmoko committed:
Created OEIMultiplier

Added a function to compute the required thrust-to-weight ratio multiplier for simulating engine inoperative conditions in novel concepts.

prmoko committed:
Created Sigmoid

Created a function to evaluate a sigmoid curve for a given set of parameters to deduce the required climb gradients under engine inoperative conditions for novel concepts.

prmoko committed:
Updated Jet25 OEI Functions

Included an option to use a novel set of performance requirements as a substitute for the existing 14 CFR 25 ones.

prmoko committed:
Updated TestConstraintDiagram

Added the necessary parameters to test creating a constraint diagram with the novel engine inoperative performance requirements.

prmoko committed:
Updated JetCeil

Instead of using the climb gradients for the enroute climb, a small climb gradient should be used to represent that the aircraft can only climb at a shallow rate to its absolute ceiling.

prmoko committed:
Created Boeing777

Created a constraint diagram for a Boeing 777 that verifies the one provided in the Aircraft Design Metabook.

prmoko committed:
Updated ConstraintDiagram

Allow turboprop and piston aircraft to also be certified by 14 CFR 25. Currently, a P/W-W/S diagram is returned if this option is selected.

prmoko committed:
Updated FaultTreeAnalysis

Reduced the calls to the Law of Absorbption to make the code run faster.

prmoko committed:
Updated FaultTreeAnalysis

Vectorized the Law of Absorbption by using a 'FailID' instead of a string to represent each failure mode.

prmoko committed:
Updated FaultTreeAnalysis

Speed up code my calling the Law of Absorbption after enumerating every 10% of combinations.

prmoko committed:
Updated Jet25_119

Replace the hard-coded landing weight fraction with a parameteric one.

prmoko committed:
Updated Boeing777

Added the CFR requirements for selecting the correct certification criteria. Added the landing weight fraction as a parameter.

prmoko committed:
Updated Jet Constraint Functions

Aside from Roskam's sizing constraints, Mattingly's constraints from 'Aircraft Engine Design' (3rd Ed.) were also included. These constraints combine linear and inverse terms of wing loading to generate a more constrained feasible region.

prmoko committed:
Updated Boeing777

Updated the Boeing 777 sizing to account for Mattingly's constraints. Added new fields to the Aircraft Structure based on the needs of Mattingly's equations.

prmoko committed:
Resolved Conflicts on Jet25_111/121a Files

Resolved conflicts from prior stashed changes.

prmoko committed:
Removed TestConstraintDiagram

Removed the test for the A320 because we have better data points to use.

prmoko committed:
Updated Jet25_111

Removed old code after resolving merge conflict.

prmoko committed:
Updated Jet25_121a

Removed old code after resolving merge conflict.

prmoko committed:
Updated Boeing777

Add cleanup code back in, allowing only one constraint diagram to be plotted before closing all figures.

prmoko committed:
Updated ConstraintDiagram

For turboprop/piston aircraft using 14 CFR 25 regulations, convert the power-weight ratio to a power loading ratio for the constraint diagram. Also, convert the wing loading to kN/m^2 instead of kg/m^2. Also, updated the index of the contour that the text should lie on.

prmoko committed:
Updated 14 CFR 25 Constraints

Added conversion from W/P to T/W for all constraints.

prmoko committed:
Updated JetLFL

Added Roskam's approximation for landing field length and added conversion from W/P to T/W.

prmoko committed:
Updated ConstraintDiagram

Removed hard-coded conversions from W/P to T/W and vice versa. Temporarily removed the service ceiling constraint.

prmoko committed:
Updated JetTOFL

Removed the unit conversion and use the takeoff speed directly.

prmoko committed:
Updated JetTOFL

Use the stall speed on takeoff.

prmoko committed:
Updated Climb Constraints

Added new requirements derived from de Vries et al., which do not use OEI correction factors. Also, a constant lift coefficient is used, with varying dynamic pressures (airspeeds), allowing all lift coefficients to be achieved with an appropriate stall margin.

prmoko committed:
Updated JetCrs

Added an extra constraint option from de Vries et al. Technically, no engine lapsing ratio was considered, but one was added to calibrate the model to the constraint diagram in literature.

prmoko committed:
Created JetDiv

Created a copy of the cruise constraint, but applied for diversions. Again, no engine lapsing ratio was considered in de Vries et al., but one was added to calibrate the model to the constraint diagram in literature.

prmoko committed:
Created JetAEOClimb

Added a constraint (as done in de Vries et al.) for AEO climb. A gradient may be specified by the user in the aircraft specification file.

prmoko committed:
Updated Boeing777

Added new requirement option and confirmed specific excess power loss.

prmoko committed:
Updated ConstraintDiagram

Modified position of text labels. Added constraints and re-ordered other ones to spread the labels out across the constraint diagram.

prmoko committed:
Created ElysianE9X

Created a model of the Elysian E9X for sizing in a constraint diagram. This may later be expanded to sizing in FAST.

prmoko committed:
Updated ConstraintDiagram

Instead of specifying the constraints in the function, define them in the Aircraft Specification File and evaluate them in this function.

prmoko committed:
Updated Aircraft with Constraints

Added relevant constraints to be included in each configuration's constraint diagram.

prmoko committed:
Created SUSAN Model

Created a model of SUSAN that can be used for making a constraint diagram.

prmoko committed:
Updated ElysianE9X

Fixed some of the aerodynamic parameters to match the assumptions made for SUSAN.

prmoko committed:
Merge branch 'main' into FaultTree

prmoko committed:
Updated SUSAN

Removed extra comments and fixed necessary values.

prmoko committed:
Deleted 14 CFR 23 Requirements

Deleted requirements that are not used in FAST. However, they are stored in a personal repository in case they are needed again.

prmoko committed:
Moved Specification Files

Created a separate package for the specification functions to reside that is separate from the constraints.

prmoko committed:
Created Contents Files

Added Contents.m files for the packages in the ConstraintDiagramPkg.

prmoko committed:
Updated Files in ConstraintSpecsPkg

Removed parameters that were not used in any of the constraint functions.

prmoko committed:
Created README

Created a README for the ConstraintDiagramPkg. This function details how to setup a Constraint Specification file and the possible parameters that could be included in it.

prmoko committed:
Removed Old Safety Analysis Files

Removed the fault tree drawing capabilities from FAST, but have them in a personal code base, if ever needed.

prmoko committed:
FTA Helper Functions out of FTA Code

Established separate function files for each helper function in the Fault Tree Analysis.

prmoko committed:
Updated FaultTreeAnalysis

Removed all helper functions and modified some of the variable names.

prmoko committed:
Updated ComponentDatabase

Updated the documentation.

prmoko committed:
Created Contents File for SafetyPkg

Created a Contents.m file to list all of the functions in the SafetyPkg.

prmoko committed:
Updated README

Fixed typo.

prmoko committed:
Updated README

Fixed typos in the ConstraintDiagramPkg README.

prmoko committed:
Updated FaultTreeAnalysis

Removed documentation for an input that is no longer needed.

prmoko committed:
Created README for SafetyPkg

The README primarily contains information about using the SafetyPkg, defining an architecture matrix, and how to accommodate time-varying failure rates into the FTA.

prmoko committed:
Updated FT Examples

Removed the old FTA examples that no longer work and moved the updated ones to a package within FAST's SafetyPkg.

prmoko committed:
Contents and README for SafetyPkg Test Cases

Added a README and Contents.m file for the TestCasesPkg that resides in the SafetyPkg. These tests can be used to confirm that the FTA is working properly and provides users with simple examples that they can learn from to construct their own architecture matrices.

prmoko committed:
Updated FaultTreeAnalysis

Allow inactive components to be included in the system architecture by checking both the inputs and outputs for determining sources and sinks.

prmoko committed:
Updated Examples for FTA

Updated the examples to be cell arrays of characters to ensure compatibility wiht GNU Octave.

prmoko committed:
Updated FTA for GNU Octave

Modified the string comparisons and representations to be compatible with GNU Octave.

prmoko committed:
Updated ComponentDatabase

Modified the string comparisons to be compatible with GNU Octave.

prmoko committed:
Updated ConstraintDiagram

If infinite contour values are returned, then set them to NaN, which allows MATLAB to save a figure.

prmoko committed:
Updated ElysianE9X

Used landing lift coefficient of 2.5 to better illustrate sizing point selection.

prmoko committed:
Updated ElysianE9X

Added info about which paper the data is from.

prmoko committed:
Updated ElysianE9X

Ensured consistency between all values used to make the constraint diagram. All values will come from the 2024 paper.

prmoko committed:
Updated ConstraintDiagram

Modified green to match the IDEAS Algae Green color.

prmoko committed:
Updated ElysianE9X Constraint Diagram

Added more detail on the plot and gridlines.

prmoko committed:
Updated ElysianE9X Constraint Diagram

Added more plotting capabilities for comparing existing and candidate design points.

prmoko committed:
Updated ElysianE9X

Removed extra plotting code.

prmoko committed:
Updated ConstraintSpecsPkg Contents

Removed function name that wasn't included in the repository.

prmoko committed:
Updated README Files in SafetyPkg

prmoko committed:
Updated FaultTreeAnalysis

Updated the documentation and error messages.

prmoko committed:
Merge branch 'FaultTree' of https://github.com/ideas-um/FAST into FaultTree

prmoko committed:
Updated Constraints and Multipliers

Updated the curve fits with the latest specific excess power loss fraction values and new constraints.

prmoko committed:
Updated ElysianE9X

Added code for plotting.

prmoko committed:
Revert "Updated ElysianE9X"

This reverts commit faebbf0265b4c8cc37ed838dd80a4f1fc5517230.

prmoko committed:
Updated ElysianE9X

Updated the Oswald efficiency factors using the corrected equation from the literature.

prmoko committed:
Updated Roskam Constraints

If using the Roskam constraints for a power-based sizing, convert the W/P to T/W by the conversion provided.

prmoko committed:
Updated Roskam's Cruise/Diversion Constraints

If performing a power-based sizing, the lapse rate is not needed. Also, cleaned up the lines of code.

prmoko committed:
Updated TOFL Constraint

Instead of operating at the stall speed, add a 20% margin (accidentally omitted previously). Also, corrected the wing loading to have units of pressure, not mass per length^2.

prmoko committed:
Updated Landing Constraint

Fixed wing loading units to be units of pressure, not mass per length^2.

prmoko committed:
Updated ConstraintDiagram

When a power loading diagram is generated, let the horizontal axis have units of kN/m^2. This was accounted for previously in the constraint diagram, but not in the axis labels.

marnson reviewed (approved):
FTA Stuff is great and super easy to understand. I left one comment on a readme that needs an entry updated. Make sure theres no large GNU snags, but if we miss something we will catch it in the GNU PR

marnson commented on +ConstraintDiagramPkg/ConstraintDiagram.m:
```diff
@@ -0,0 +1,196 @@
+aclass = Aircraft.Specs.TLAR.Class;
+
+% get the certification basis
+CFRPart = Aircraft.Specs.TLAR.CFRPart;
```
nevermind i see how things were created. just want to make sure nothing conflicts with the aircraft sizing stuff

prmoko committed:
Updated Constraint Diagram

Added semi-transparent minor gridlines to make the plot easier to read.

prmoko committed:
Merge branch 'main' into FaultTree

prmoko committed:
Updated CreateCutSets

Modified the explanation of what a cut set is and how they are identified.

prmoko committed:
Updated Contents in SafetyPkg

Added the 'TestCasesPkg' folder in the list of contents.

prmoko committed:
Updated README

Updated the README files to include the '+ConstraintDiagramPkg' and '+SafetyPkg', along with brief explanations of both packages.

prmoko committed:
Updated CHANGELOG

Updated the CHANGELOG to reflect the latest changes and introduce FAST Version 0.4.0.

prmoko committed:
Updated from Main

prmoko committed:
Updated OEIMultiplier

Utilizing a new function after re-analyzing new data.

prmoko committed:
Updated Sigmoid Curve Parameters

Updated the Sigmoid Curve parameters after re-fitting the curves for the updated specific excess power loss fractions.

prmoko committed:
Updated JetTOFL

Takeoff speed should be 1.1-times larger than the stall speed, not 1.2-times larger.

prmoko committed:
Updated Jet-based Constraints

Included only correction factors for OEI and MTOW corrections, matching the E9X sizing process.

prmoko committed:
Updated E9X Specs

Added AEO constraint back into the sizing process.

prmoko committed:
Updated Jet-based Constraints

Added the correction factor back in.

prmoko committed:
Updated Elysian E9X

Use the landing lift coefficient of 2.5 for the diagram (but 3.1 in the landing constraints).

prmoko committed:
Updated Jet-based Constraints

Include all correction factors in the analysis.

prmoko committed:
Updated Elysian E9X

Set the MCT and Temperature corrections factors to 1, as they are not used in the E9X analysis.

prmoko committed:
Updated Jet25_121d

Removed hard-coding.

prmoko committed:
Updated ElysianE9X

Cleaned up code.

prmoko committed:
Merge branch 'main' into FaultTree
